lib/Dialect/CGGI/Transforms/BooleanLineVectorizer.cpp

```diff
@@ -30,15 +42,126 @@ namespace cggi {
 bool areCompatibleBool(Operation *lhs, Operation *rhs) {
   if (lhs->getDialect() != rhs->getDialect() ||
       lhs->getResultTypes() != rhs->getResultTypes() ||
-      lhs->getAttrs() != rhs->getAttrs()) {
+      lhs->getNumOperands() != rhs->getNumOperands() ||
+      // Attributes do not need to match for Lut3
+      (!llvm::isa<cggi::Lut3Op>(lhs) && lhs->getAttrs() != rhs->getAttrs())) {
     return false;
   }
   // TODO: Check if can be made better with a BooleanPackableGate trait
   // on the CGGI_BinaryGateOp's?
   return OpTrait::hasElementwiseMappableTraits(lhs);
 }
 
-bool tryBoolVectorizeBlock(Block *block, MLIRContext &context) {
+FailureOr<SmallVector<Attribute>> BuildGateOperands(
+    const SmallVector<Operation *> &bucket, MLIRContext &context) {
+  SmallVector<Attribute> vectorizedGateOperands;
+  for (auto *op : bucket) {
+    FailureOr<Attribute> attr =
+        llvm::TypeSwitch<Operation &, FailureOr<Attribute>>(*op)
+            .Case<cggi::AndOp>(
+                [&](auto op) { return StringAttr::get(&context, "and"); })
+            .Case<cggi::NandOp>(
+                [&](auto op) { return StringAttr::get(&context, "nand"); })
+            .Case<cggi::XorOp>(
+                [&](auto op) { return StringAttr::get(&context, "xor"); })
+            .Case<cggi::XNorOp>(
+                [&](auto op) { return StringAttr::get(&context, "xnor"); })
+            .Case<cggi::Lut3Op>(
+                [&](cggi::Lut3Op op) { return op.getLookupTable(); })
+            .Default([&](Operation &op) -> FailureOr<Attribute> {
+              op.emitOpError("Unable to find printer for op");
+              return failure();
+            });
+    if (failed(attr)) {
+      op->emitOpError("Failed attr");
+      return failure();
+    }
```

Comment:
```suggestion
              return failure();
            });
    if (failed(attr)) {
      op->emitOpError("unsupported operation for vectorization");
      return failure();
    }
```

(I don't think it's necessary to emit an error twice on the operation)

Reply:
Done!